libraries/chain/include/eosio/chain/apply_context.hpp

```diff
@@ -134,54 +134,152 @@ class apply_context {
                if( !tab ) return -1;
 
                const auto* obj = context.db.find<ObjectType, contracts::by_secondary>( boost::make_tuple( tab->id, secondary ) );
                if( !obj ) return -1;
 
+               primary = obj->primary_key;
+               secondary = obj->secondary_key;
+
                itr_cache.cache_table( *tab );
                return itr_cache.add( *obj );
             }
 
-            int lowerbound_secondary( uint64_t code, uint64_t scope, uint64_t table, const secondary_key_type& secondary ) {
-               auto tab = context.find_table( scope, context.receiver, table );
+            int lowerbound_secondary( uint64_t code, uint64_t scope, uint64_t table, secondary_key_type& secondary, uint64_t& primary ) {
+               auto tab = context.find_table( context.receiver, scope, table );
                if( !tab ) return -1;
 
-               const auto& idx = context.db.get_index< chainbase::get_index_type<ObjectType>::type, contracts::by_secondary >();
+               const auto& idx = context.db.get_index< typename chainbase::get_index_type<ObjectType>::type, contracts::by_secondary >();
                auto itr = idx.lower_bound( boost::make_tuple( tab->id, secondary ) );
                if( itr == idx.end() ) return -1;
                if( itr->t_id != tab->id ) return -1;
 
+               primary = itr->primary_key;
+               secondary = itr->secondary_key;
+
                itr_cache.cache_table( *tab );
                return itr_cache.add( *itr );
             }
 
-            int upperbound_secondary( uint64_t code, uint64_t scope, uint64_t table, const secondary_key_type& secondary ) {
-               auto tab = context.find_table( scope, context.receiver, table );
+            int upperbound_secondary( uint64_t code, uint64_t scope, uint64_t table, secondary_key_type& secondary, uint64_t& primary ) {
+               auto tab = context.find_table( context.receiver, scope, table );
                if( !tab ) return -1;
 
-               const auto& idx = context.db.get_index< chainbase::get_index_type<ObjectType>::type, contracts::by_secondary >();
+               const auto& idx = context.db.get_index< typename chainbase::get_index_type<ObjectType>::type, contracts::by_secondary >();
                auto itr = idx.upper_bound( boost::make_tuple( tab->id, secondary ) );
                if( itr == idx.end() ) return -1;
                if( itr->t_id != tab->id ) return -1;
 
+               primary = itr->primary_key;
+               secondary = itr->secondary_key;
+
                itr_cache.cache_table( *tab );
                return itr_cache.add( *itr );
             }
 
-            int find_primary( uint64_t code, uint64_t scope, uint64_t table, uint64_t primary ) {
-               auto tab = context.find_table( scope, context.receiver, table );
+            int next_secondary( int iterator, uint64_t& primary ) {
```

Comment:
Since we are iterating until the end of the table, not the end of a particular secondary key value, shouldn't this emit the value of the secondary key too (to keep convention with the other iterator operations)?